types/react/index.d.ts

```diff
@@ -981,20 +981,16 @@ declare namespace React {
      * @see https://reactjs.org/docs/hooks-reference.html#usereducer
      */
 
-    // I'm not sure if I keep this 2-ary or if I make it (2,3)-ary; it's currently (2,3)-ary.
-    // The Flow types do have an overload for 3-ary invocation with undefined initializer.
-
     // NOTE: without the ReducerState indirection, TypeScript would reduce S to be the most common
     // supertype between the reducer's return type and the initialState (or the initializer's return type),
     // which would prevent autocompletion from ever working.
 
-    // TODO: double-check if this weird overload logic is necessary. It is possible it's either a bug
-    // in older versions, or a regression in newer versions of the typescript completion service.
-    function useReducer<R extends Reducer<any, any>>(
-        reducer: R,
-        initialState: ReducerState<R>,
-        initializer?: undefined
-    ): [ReducerState<R>, Dispatch<ReducerAction<R>>];
+    // NOTE 2: The past concern around autocompletion with the returned state object
+    // seems to no longer be an issue. See https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/63607
```

Comment:
We can remove NOTE and NOTE 2 now. 

I can't even get autocomplete to work with the previous typings (down to TS 3.3): [Playground with previous types](https://www.typescriptlang.org/play?ts=3.9.7#code/JYWwDg9gTgLgBAbwK4GcCmAlNATJBjNKAXzgDMoIQ4ByKNAQzxmoCgWB6duGNFGF4ADseUUozRwAgmDABlGPR6IWcOIPog0ALjh8oQgOYBuFXHoHtapCABGhE0TYwAnmAnSwkpsAiCUcAF5TAB9EbldLACILGAB5ABtsQki4R1VQhHC3HUi6dBhJCxSiExZSJEFvXzg6XAIoAAoUHQ95RTQAGjMqwRaZLxgfPwBKPrkFJQRTFAB3YBg8AAs4BsZB3wA6Fzdh5VVVPHp0OGi0OMTkrVN9mrOkKEEwjeeULvNLFA33uABqOABGVImG6HY65XhnQpoSJXG6qOgwe6PTLPDavMwWHQABiBpkcjjYqEwOHwhAatVJUC6mXeWgALAAmLqpYZGIA)

Proposed typings don't have autocomplete either: [Playground with proposed types](https://www.typescriptlang.org/play?#code/PTAECcFMEMGMBcB0ATR8DOAoTJTwBYCW6eks+AdobNADa0CeoyA9pCXLJAA7yjSh0ZFhWT9wAcwCuAW0gV4AGlAAjKX0LwA5B1pRoyJlNGRw-Zjyg14hEaADu0cFQoScYaKNCadoCiz5IADd5CEhaGCExdBZQbnAWFWgVRlVIeHhTPFj-PmRIADNCCkhvJEx4Bm5SgBFibmh4cgAeAEEAPlAAXlAACiC6KUgALlBWgEpuzqCWQmQAbndQAGVirn4EWwpQAH0aCh200GN8opLkZWR6xvJQGWgmFVKaekgxe018FnV+CiYG8DQOSZcDoRAVKq1a5NfAAdU+33grU2dh6vUmXWmswWSwAqhRaIQANalKDIKQAD2UBFKcBsIhIe08hyevwYBGKbkq1VAACU3lIuOBmstlB1un14sFlvBGiMVso6VtRhMpitFrgAJIFY5CfnkoUbLi8DhhA1ZD4EREbekUS7Q273R7POgRd4In6ef5OIHpUxgiE8-WC0zwq3qZG2kWdNFSoIyuWjZYYzrLDVgbnsXVvbJ4cBMTxiOPyPg00CwFgybiECJmAoJGRmwWclZLeCxAToKTVcCZ8uVmR2S34PD4UpCCteJzSOQKXNSPUCoXowOlYNChOZZq80CQCmZUQkdemZpexV-doxvm7-fyZBHpcn4oFLKitmdAD8K1AoxKIXAix9se4CRls243ge958o+wpnu+V47nukEPuasF-Moz5ZOKX6tD+fjBKY6agAAKmO3j5AomhMOQZBEt4JCsCUDifKOpTLL6ABqgy0rQEgsOAnyNr0ADyKgAFZkEgxDjMo9hELcxCCPAAkIFkBCeN0XRdKu0GoZukBhl8EYohQ4FIXeKEhuAhmIqBIinheMaYKA17mYeulWTZxlRphZjLJ+36-gRAHYMAABUYXOS5LlhWM2x0CCFCNIQIS5gABgukD6Wl4LRTFUXRbFGWLqhaX0bqUiuv8UAvoCKSlO2oDFVlsqZGVcmhAw3ygPg0CpRWVYRBSSlyqAtAsBI1Cjo03gUDMtAhFgeWgLFMhSLQNjcBEghSCoAC0Ay0EMYKgJqfB0DEY3pCQXVSKALC8IQMiEAAXqUPYFPx9wUOsn1mAN3AiCWJDqaWAkSBIWT5DwBUxcc3DIHKJBPDQmWgLd5YaQ06AkGlVzoA0MJlaw9jbMU6CZAY906i8tBJLARIBstkXM6AAAC-7oFsoAAIwAGyIAAHIgAAMsMrezQilPgGTcOgwwgPoCBiWC-ESMArCwOgwBfCwjN7TVpjyFwiAyzItAAMSZWSVni2FwBLCw-7jVTHVQMwDojhW61iJwPB8CL4gzsD4L5LAtBOKUBTGCZ2bAWZt7ucBXlIiZDkMO0yiau0vRRTbQqjLyihRcUmiEHQr2mK0kijJqxcuaXNgV294CjL0061ymHkbq1BkeqnUa8u0mDjKMADawH6SndmmUPyh1ATNxwv3M8ALoahFzmxa08UbaYSU2KljXNdluUs0VmXAWVikLpV9DVYUpjJNtx+ZdlDhjtsGO9f1lZbXuI1MhjQmlNUGs15qLS3ncdam1tpdn2odY6iBTrnVoJdCIGB0bdQejYZ6b04imD+t9X6-F+xViBgoEGvUwaEAhlDSAMMJbdkRpkZGZBoBowxvsOI0AcZNXxoTcgxMWCk1mhTGAYgWA01dPTRmZ8oEc39NzfmQtRYKKlj1WW8tFYwGVqrSQGsWBax1iwPW6ADaPygD9SApt4DmytkIfOpgt4O1wE7UwLt3hjndgIpe-YfZGn9qAQO05ZAh0wGHCO7to4-VtHHGCCdkLd1DCvNOXpLy5xck41ufJ66zTLs3KuNdkngCnqkwemcS5UCboSFuH5RgnEKMUN4I9x6T17tPNOc9QAL0EcvcMA8tjrxwJvCWO9+B72cMlI+sQT69xylvKBzUr7lVvlVOIhs6ov1mW-eZH9OrdR-s8P+Q1AGlHGpNWA00NBzRYAtdgUC1obUIP-HaCDuInTOhM9B10sF3RwU9SuBDwBEM8CQ-6f8KGYLAcpWhkMzDQ24FA5hSM0ioyEH8zG2xsa418UTZgIiyYUHEVTKRmN6CyKZufdmnNlEC2FmLCWbMNEy3gHLBWwAlbwBVogNWhjjG631ps42NizaW2tjBFxjtnYsFdt40oAAiTUCq7gPCOJ2XaQg+BkvaXKbc7R5jZjEI1eITs5i0nUEYk56Qtjgi1DqJVKr7gwizLqrcQ8bx0lSDSbYZYI4U3ujK12NZaATPGvYfJNTK5mEaqySsmhMioDbGRf12qg1iEKC+TYIRUhPCORwEghAdRlnNZRSooAYmx16MNTEoAKSTEUiW6p5dammFDmQKJUcY5xMvgkxCicoLxzgukrOOc84wULnkxuLbo3VwkLXUAAAyEp+l9VTubYUnJ7dimaiXSu3u+qu5usgPq1poAJ4wVXd03pS9twwRnvq9owzwrUvGQlfe0yGo7KEKfRZEtlkwWviQNZ98NmWOfl+pquy5TtU-pio5ZD-7DQpqNC5oDqHgLuZAiWTzYHjl2gdD5yCvkXViBgm62DHp4PeoQr6YKo6kIBlCqhM1YV0IRQwpFTCEaopRhwjFXCsa8NxZ7YRojyaU0kdIilcA5F-tioo0EdLVGMoUyyrRHKuU8r5ZrbWgrzHCusbY+xErUJSrcemitpD6wMNAI6w1rpQB8SzLw4Brh+CFo3a2iFc1TCYM0LNRqCrj36oVe28OkcK3du5r21CiSLIlPTueDOo7MlNgLrkqpBTvNztruu7L0a24d1Okey9B6h5novXpcrmcemezvahB9Q8n0b1fbvRKn70rQbavI-9sWrJAYqusqUtUINdZ-XsjqX9Dl9WOYNABKGgFoauWA4oECHk4ZgS8uBBHEHsGI6gn5mCMYAuo8C0F1irMQvISUSh1y8xwvoYw2KKLWFov46UQT2LhP8NEwS8TxLJPU3JXTWTVKFG0rsCohl6jbOsvZTouk2mDG6ZMWYixtURXGfFY4yVK0HZLAAHLCWIgAUVGMOa0ZZj2zSuFYW0yhiKQmWLAASvAHDfFoGIJx35Y0NTIoOANA1BwUCWF2HsfYnjwHsAw31ZFsm+CgPAKQzg8CQl+Masi066D6T6KQptBWW6K-SCr31kIZJLDktQEc9hOdFigCEOcHD2wAwwdzeslZdz-g5+AIknJwRE5J6T0AAAmUYpF3q8L4JOIU8UEgnH4Ja13NqhwsTLEr03OZFulESBJBAYuGEyBBjkMjIh4XqrJjjIYyDlhw40yASaBBdqIAGsAGoTTqnhAYEz6oyA28d80F3nvbwNbEFgAuLmDJgB8wAMx8xFgAdgiR2yLlae0lSsiKMUY6skTr6PLBU-AVRd1FFlmp+kkyVbfO3Eyx+1QzDmMM0KGZ2DwEwMUEEBQ4ClFaNwbguuABvKKJKOQUYCmASVwRYFyaASGX8WQJ4EKAAX2wD7F-24BnhIG0hcgAB9QAAC1dqhRgFVIZ4BhIudTAVVkCcC8CCD5QFUoAtVWhIZKDFhMA19uZslegD80D9JFRb8xg-8MDR4BD-9e48Cop0BLRbgb9bQ0BzdxC8oaAMViD0gyD8hwAFVhhxYwhldVd8DEADD0BFRYDBBEAYDSgABqXmUARAqA6KJQxVBg9IJgyATQ7QjPPQ0AAwxAIw-gEwwOWwqKZA5A7AfrZcbJZQfA8w0YAAFhD2UEQPGHmCAA)

Reply:
Strange. It works in the StackBlitz I made. If you go to line 40 and erase the `count` property and start typing again it autocompletes it. https://stackblitz.com/edit/react-ts-gy5gv6?file=App.tsx

Anyway, pipeline failed so I'm exploring that. And I'll erase those notes

Reply:
My overload is being used but the other specific overloads are causing my overload to not be able to infer the type. I will explore this more.

Reply:
@eps1lon  I came up with something that might work. 2 overloads instead of 6. I just updated this PR.

Tests pass locally _except_ for the test I added in this PR. I don't know why, because in VSCode running 5.0.0-dev20221218 It's correctly inferring how I want so the test should pass, and in the TS environment like what you shared, it's inferring correctly with v4.9.5. If it's failing for me locally, I bet it will fail in the pipeline. If it does fail, should I just remove my added test? It will pass existing tests and apparently _sometimes_ be better than what's there now. 

In VSCode it's autocompleting, and I just confirmed it is in [the TS environment](https://www.typescriptlang.org/play?#code/PTAECcFMEMGMBcB0ATR8DOAoTJTwBYCW6eks+AdobNADa0CeoyA9pCXLJAA7yjSh0ZFhWT9wAcwCuAW0gV4AGlAAjKX0LwA5B1pRoyJlNGRw-Zjyg14hEaADu0cFQoScYaKNCadoCiz5IADd5CEhaGCExdBZQbnAWFWgVRlVIeHhTPFj-PmRIADNCCkhvJEx4Bm5SgBFibmh4cgAeAEEAPlAAXlAACiC6KUgALlBWgEpuzqCWQmQAbndQAGVirn4EWwpQAH0aCh200GN8opLkZWR6xvJQGWgmFVKaekgxe018FnV+CiYG8DQOSZcDoRAVKq1a5NfAAdU+33grU2dh6vUmXWmswWSwAqhRaIQANalKDIKQAD2UBFKcBsIhIe08hyevwYBGKbkq1TGKIodXQDRhzQASm8pFxwGL0IEKZlRBw-gBtAC6nS6mFAWtAYvJkulsvlyBISs8DBVoAA-KABUKWrqJaYDUqAAxq0CjW03OEI9TI+kURbYYAAKhDmpDY22dBBFEahBC2VAAAMpEJlvBGpBk+DQOG8ym05AHZLk94SGmpHRUvFCqZkhEk6n05nMmX7PhQgxvqB8NBE7AWDJuBEKYJW6VaCwJNQ8H2NBQZrQQlgCzIpLQbCPSugpCoALQDWhDMGgACSfDoMVAEQwoG7UlALF4hBkhAAXqVquACixwPcKHWX8zEHYcRHkO8CEaPBwEICQJCyfIeAjY5uGQLMSCeGgi3vHt9jiaB0BIZMrkFb0y1YextmKGUYDEFgClAF5aCSWAiTBCMUIAARCUEtlAABGAA2RAAA5EBdbihFKfAMm4dBhhAfQEAAKzBP8JGAVhYHQYAvhYdj9ygApTHkLhEFkmRaAAYiLMlHXACNgEwfJYFoJxSgKYw+WOIQS1MZplmUVpQEgOV5GNNlVXaXpNW1ezJVGXpayCDMs1GILQEQbKnAkBSxgxTogrirVik0Qg6DSzIMswcZRiVTL-S2L1hQ6UAVUWUN80jVpo03Uw4xsRN4FiZtICq7Nc26ws-PFUty18qt6H+Yz6xSUoRpm8aJ3bTttgfXt+2eIdtzHGUsxvadZyghclxXFD103Qht0EPdD0GdhEHPS9aGvW8SAO58bDfT84lMYCAKAv8mJO8CFBIG6YLghCzCQ7gUKkNCMLSbChFwx98IaIiU1Iu18AolgqO8ChaIMJ9GOY1j2Km7jePQfjhLEiSpMgGS5IUpSYFU9TJC0lgdL0lgDPQIy6ygQDIAs+ArNsoQEtMJyXLIdyoFALzAIDXzizmgLMrPYLQvChUorVWLtTCPVTCSlKJoy5RssQXL8omKYVkUErqfKuhWkkUZzYDsr4CSwgw8KlZavqxq+RaloOg67AlkyGVMGKEECjgUpWm4bgJtAABvOK4zkUYZVg1xFi1aAENGChZCecBFgAX2wblC+LpqGW6OKAB9y7wSFRgAIgQ+AAHlaHycBJ9AbutVHsvx+qKeoCEJEEOXzug0wfWfPV8Bem94uJuUOktlGIvuAHmm6rGK+J3LuL0A+GE+lvkQ0EhJMCu9saB42nukeei9J7DADlqKA8ApDODHh7dAN9m6CE9ghUAABqQSK8G7alAaUSeO90itH3jA+2cD0iIO2BvFBaCRigBdPguK3du5LAAAoJCCHMUokcKq0HHFmZQxQTJmDPsIzIm9ICYCLP5c+cUT6GzPhfF+D9S7AKofA2hyDsqoP4OgsETdSi4IEqwrUnd-Zag3iY0YAAWAATMoKuTCtA1E8IQcIWhu7jCDIOGmfAlTnUyBaHo8iTaKK1BfG+fJ779z5OgdRb8LqYg-vbL+mhbi9D-hQAB1QgGwKYoRYhM9IGmGgUU6hCCkH0P0YwmumDTF4MPkUohoASHsDIRQqpYQal0KyvUwxTCWGtPtqvFe1ix52NAE4lxQI3EeKoN4letUgwhMViY4MYAAAqkIHYOSkRtSEN91DixOrefhVAbB0COXI2ajsol9EviXCcyTXmpM6Fo+2OjamDMQAYmZxisFmIsZMuKtj0FzL8As0Y7jPHeN8f4hkQSNmOLCUbBRdsDmSimZCph0LXFwqWV42gWhVl+OwBsppQA) too.

Reply:
Less overloads is always a win. Especially if we can have 3 or less overloads since that has poor debuggability: https://github.com/microsoft/TypeScript/issues/52745

Reply:
Yes it failed in the pipeline for some reason. I'll try to figure it out a bit longer, but I could just remove my test and make this a "simplify useReducer overloads" PR (maybe it fixes autocomplete too) and in every environment except whatever this specific one is, it will infer the reducer state like I wanted. 
![image](https://user-images.githubusercontent.com/13700926/219875488-31331de6-6043-4bb5-b271-c1dfe24eed1c.png)

Reply:
I got it passing everything locally! `$ExpectType` is very picky and not helpful at telling you how you're using it wrong. But now I know.